### DISCUSSION TITLE
Initial consensus item ordering latency metrics

### DISCUSSION THREAD

**Discussion by dpc** [fedimint/fedimint#6521] [General]
Now that we've deployed 0.5 to signet demo federation, we can start checking consensus ordering latency from https://github.com/fedimint/fedimint/pull/5747 . (Ping @joschisan who came up with it)

```
[root@fedimintd-01:~]# curl -s localhost:8175/metrics | grep consensus_ordering_latency_seconds
# HELP fm_consensus_ordering_latency_seconds Duration of ordering a batch of consensus items
# TYPE fm_consensus_ordering_latency_seconds histogram
fm_consensus_ordering_latency_seconds_bucket{le="0.005"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.01"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.025"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.05"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.1"} 19
fm_consensus_ordering_latency_seconds_bucket{le="0.25"} 179189
fm_consensus_ordering_latency_seconds_bucket{le="0.5"} 251727
fm_consensus_ordering_latency_seconds_bucket{le="1"} 252711
fm_consensus_ordering_latency_seconds_bucket{le="2.5"} 252959
fm_consensus_ordering_latency_seconds_bucket{le="5"} 252974
fm_consensus_ordering_latency_seconds_bucket{le="10"} 252979
fm_consensus_ordering_latency_seconds_bucket{le="+Inf"} 252988
fm_consensus_ordering_latency_seconds_sum 58929.05764798793
fm_consensus_ordering_latency_seconds_count 252988
```

Just to explain reading result of prometheus histogram: around half (`179189`) of the items had latency less than 0.25s, and almost all (`251727`), less than  0.5s (this includes the items that were <0.25s already).

Seems like what we would expect - this Federation is just 4 VMs in the same Hetzner datacenter. Looking forward to seeing this metric from more less than ideal deployments.

**Comment by joschisan** [comment]
This is very much what we would expect.

**Reply by elsirion** [reply]
I remember us being surprised about high LNv1 latency, did you ever benchmark LNv2 against it? At least we can rule out consensus it seems.

**Comment by elsirion** [comment]
Some additional data from early v0.5 testers running a federation with guardian nodes in multiple different DCs:

```
fm_consensus_ordering_latency_seconds_bucket{le="0.005"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.01"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.025"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.05"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.1"} 0
fm_consensus_ordering_latency_seconds_bucket{le="0.25"} 18868
fm_consensus_ordering_latency_seconds_bucket{le="0.5"} 54826
fm_consensus_ordering_latency_seconds_bucket{le="1"} 57336
fm_consensus_ordering_latency_seconds_bucket{le="2.5"} 57422
fm_consensus_ordering_latency_seconds_bucket{le="5"} 57472
fm_consensus_ordering_latency_seconds_bucket{le="10"} 57474
fm_consensus_ordering_latency_seconds_bucket{le="+Inf"} 57474
fm_consensus_ordering_latency_seconds_sum 16797.331111376978
fm_consensus_ordering_latency_seconds_count 57474
```

Very similar numbers to what @dpc saw in a single-DC deployment, I guess the single-DC one was limited by the minimum time between units @joschisan while the latency in the real world one is only slightly higher than the "simulated latency" in the single-DC case?

**Reply by joschisan** [reply]
Exactly. By the way I find the granularity of the buckets a bit high. I would prefer 0.1s buckets under 1s and then 0.25s und 2s.

**Reply by dpc** [reply]
Was thinking the same thing today. :D

**Reply by dpc** [reply]
https://github.com/fedimint/fedimint/blob/840f993805754c8ec7199db19a02fc33ffe8eb83/fedimint-server/src/metrics.rs#L82 , we can add another argument with the buckets like here: https://github.com/fedimint/fedimint/blob/840f993805754c8ec7199db19a02fc33ffe8eb83/fedimint-server/src/metrics.rs#L27
